### PR TITLE
Spike for encrypting configuration cache data

### PR DESCRIPTION
**Pull request by abstratt**
This is a exploratory spike to assess the impact of implementing encryption of configuration cache data.

Issue: #23866

### CODE REVIEW

**Comment by abstratt** [comment]
@bot-gradle test ACC

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Combinators.kt [hunk 1]
```diff
@@ -346,8 +346,9 @@ fun Encoder.writeFile(file: File?) {
 
 
 internal
-fun Decoder.readFile(): File =
-    BaseSerializerFactory.FILE_SERIALIZER.read(this)
+fun Decoder.readFile(): File {
```
Comment (abstratt) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Combinators.kt:349]: will revert

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt [hunk 2]
```diff
@@ -84,14 +84,15 @@ import java.io.OutputStream
 
 
 internal
-enum class StateType {
-    Work, Model, Entry, BuildFingerprint, ProjectFingerprint, IntermediateModels, ProjectMetadata
+enum class StateType(val encrypted: Boolean = false) {
```
Comment (abstratt) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt:87]: s/encryptedf/encryptable/

Also, have yet to test which set of state files should really be encryptable

**Comment by abstratt** [comment]
@bot-gradle help

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by cloudshiftchris** [comment]
Encrypting the CC is not a great option for security, creating additional challenges.  Discerning consumers will not want their secrets stored at all (they should be externalized, which Gradle has been touting as a best practice), or they'll want to control the encryption algorithm, keys, etc. (and you'll *still* need to secure the key).  Many consumers these days will still consider a loss of encrypted data as a breach (and, really, storing secrets on a filesystem is so 1995...).

It also doesn't address many other security challenges, such as random plugins having access to all properties and environment variables (including those that contain secrets, as there is no distinction).

A cleaner approach is first-class support for credentials; some thoughts on that [here](https://github.com/gradle/gradle/issues/22618).

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@cloudshiftchris Thanks for chiming in. I failed to provide proper context for this PR, which is not meant as a solution for #22618. I added a new issue (#23866) to provide better context for this PR.

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle cancel

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by abstratt** [comment]
Implemented in https://github.com/gradle/gradle/pull/24150

**Pull request closed without merging**